core/src/test/java/org/elasticsearch/gateway/GatewayMetaStateTests.java

```diff
@@ -245,4 +259,193 @@ public void testWriteClosedIndex() throws Exception {
+        MetaDataUpgrader metaDataUpgrader = new MetaDataUpgrader(Collections.emptyList());
+        MetaData upgrade = GatewayMetaState.upgradeMetaData(metaData, new MockMetaDataIndexUpgradeService(true), metaDataUpgrader);
+        assertTrue(upgrade != metaData);
+        assertTrue(MetaData.isGlobalStateEquals(upgrade, metaData));
+        for (IndexMetaData indexMetaData : upgrade) {
+            assertFalse(metaData.hasIndexMetaData(indexMetaData));
+        }
+    }
+
+    public void testCustomMetaDataNoChange() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        MetaDataUpgrader metaDataUpgrader = new MetaDataUpgrader(Collections.singletonList(ArrayList::new));
+        MetaData upgrade = GatewayMetaState.upgradeMetaData(metaData, new MockMetaDataIndexUpgradeService(false), metaDataUpgrader);
+        assertTrue(upgrade == metaData);
+        assertTrue(MetaData.isGlobalStateEquals(upgrade, metaData));
+        for (IndexMetaData indexMetaData : upgrade) {
+            assertTrue(metaData.hasIndexMetaData(indexMetaData));
+        }
+    }
+
+    private static class MockMetaDataIndexUpgradeService extends MetaDataIndexUpgradeService {
+        private final boolean upgrade;
+
+        public MockMetaDataIndexUpgradeService(boolean upgrade) {
+            super(Settings.EMPTY, null, null);
+            this.upgrade = upgrade;
+        }
+        @Override
+        public IndexMetaData upgradeIndexMetaData(IndexMetaData indexMetaData) {
+            return upgrade ? IndexMetaData.builder(indexMetaData).build() : indexMetaData;
+        }
+    }
+
+    private static class CustomMetaData1 extends TestCustomMetaData {
+        public static final String TYPE = "custom_md_1";
+
+        protected CustomMetaData1(String data) {
+            super(data);
+        }
+
+        @Override
+        protected TestCustomMetaData newTestCustomMetaData(String data) {
+            return new CustomMetaData1(data);
+        }
+
+        @Override
+        public String type() {
+            return TYPE;
+        }
+
+        @Override
+        public EnumSet<MetaData.XContentContext> context() {
+            return EnumSet.of(MetaData.XContentContext.GATEWAY);
+        }
+    }
+
+    private static class CustomMetaData2 extends TestCustomMetaData {
+        public static final String TYPE = "custom_md_2";
+
+        protected CustomMetaData2(String data) {
+            super(data);
+        }
+
+        @Override
+        protected TestCustomMetaData newTestCustomMetaData(String data) {
+            return new CustomMetaData2(data);
+        }
+
+        @Override
+        public String type() {
+            return TYPE;
+        }
+
+        @Override
+        public EnumSet<MetaData.XContentContext> context() {
+            return EnumSet.of(MetaData.XContentContext.GATEWAY);
+        }
+    }
+
+    private static MetaData randomMetaData(int nIndex) {
```

Comment:
Instead of randomizing, can we have explicit tests for adding, replacing and removing custom metadata?